Rubberduck.CodeAnalysis/Inspections/Concrete/IllegalAnnotationInspection.cs

```diff
@@ -1,175 +1,112 @@
+                .ToList();
+        }
 
-                var isModuleAnnotatedForMemberAnnotation = isMemberAnnotation
-                    && (_currentScopeDeclaration?.DeclarationType.HasFlag(DeclarationType.Module) ?? false);
+        private static ICollection<IAnnotation> NonModuleAnnotationsOnModules(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => declaration.DeclarationType.HasFlag(DeclarationType.Module))
+                .SelectMany(moduleDeclaration => moduleDeclaration.Annotations)
+                .Where(annotation => !annotation.AnnotationType.HasFlag(AnnotationType.ModuleAnnotation))
+                .ToList();
+        }
 
-                var isMemberAnnotatedForModuleAnnotation = isModuleAnnotation 
-                    && (_currentScopeDeclaration?.DeclarationType.HasFlag(DeclarationType.Member) ?? false);
+        private static ICollection<IAnnotation> NonMemberAnnotationsOnMembers(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => declaration.DeclarationType.HasFlag(DeclarationType.Member))
+                .SelectMany(member => member.Annotations)
+                .Where(annotation => !annotation.AnnotationType.HasFlag(AnnotationType.MemberAnnotation))
+                .ToList();
+        }
 
-                var isIllegal = !(isMemberAnnotation && moduleHasMembers && !_isFirstMemberProcessed) &&
-                                (isMemberAnnotatedForModuleAnnotation || isModuleAnnotatedForMemberAnnotation);
+        private static ICollection<IAnnotation> NonVariableAnnotationsOnVariables(IEnumerable<Declaration> userDeclarations)
+        {
+            return userDeclarations
+                .Where(declaration => VariableAnnotationDeclarationTypes.Contains(declaration.DeclarationType))
+                .SelectMany(declaration => declaration.Annotations)
+                .Where(annotation => !annotation.AnnotationType.HasFlag(AnnotationType.VariableAnnotation))
+                .ToList();
+        }
 
-                if (isIllegal)
-                {
-                    _contexts.Add(new QualifiedContext<ParserRuleContext>(CurrentModuleName, context));
-                }
-            }
+        private static readonly HashSet<DeclarationType> VariableAnnotationDeclarationTypes = new HashSet<DeclarationType>()
+        {
+            DeclarationType.Variable,
+            DeclarationType.Control,
+            DeclarationType.Constant,
+            DeclarationType.Enumeration,
+            DeclarationType.EnumerationMember,
+            DeclarationType.UserDefinedType,
+            DeclarationType.UserDefinedType,
+            DeclarationType.UserDefinedTypeMember
+        };
+
+        private static ICollection<IAnnotation> NoneGeneralAnnotationsWhereNoOtherAnnotationsBelong(IEnumerable<Declaration> userDeclarations)
```

Comment:
`None` -> `Non`

Reply:
I don't this method is in line with the definition of `GeneralAnnotation`. Can you please explain what it does, because I don't understand.

Reply:
Arrgh, I forgot to add the last condition that the annotation does not have the general flag.

This also needs a test.